File: android/app/src/main/java/com/whowouldin/whowouldwin/vm/AnimalPickerViewModel.kt
package com.whowouldin.whowouldwin.vm

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import com.whowouldin.whowouldwin.data.Animals
import com.whowouldin.whowouldwin.model.Animal
import com.whowouldin.whowouldwin.model.AnimalCategory
import com.whowouldin.whowouldwin.model.BattleEnvironment
import com.whowouldin.whowouldwin.service.AnimalImageService
import com.whowouldin.whowouldwin.service.CheatState
import com.whowouldin.whowouldwin.service.ContentFilter
import com.whowouldin.whowouldwin.service.UserPrefs
import kotlinx.coroutines.FlowPreview
import kotlinx.coroutines.Job
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.debounce
import kotlinx.coroutines.flow.onEach
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

/**
 * Port of iOS `AnimalPickerViewModel.swift` to Jetpack Compose / Android.
 *
 * Exposes a single [UiState] [StateFlow] consumed by [AnimalPickerScreen]. All
 * derived state (filtered list, canFight, lockedAnimal, customAnimal) lives on
 * the state object so the Compose side can `collectAsStateWithLifecycle()`
 * once and read everything.
 *
 * Design notes:
 *  - Cheat-code tap counter is session-only and lives on [CheatState] (see the
 *    original iOS `CheatState.shared.olympusUnlocked`). The view owns its own
 *    `olympusCheatStep` because it's a transient UI-level tap sequence.
 *  - Dual-path custom-creature flow is modeled as two VM entry points:
 *    [payCoinsForCustom] and [watchAdForCustom] (the caller hands in a
 *    "granted" result for the ad path). Both delegate to [selectAnimal].
 *  - First-free custom battle is a Boolean persisted through [UserPrefs]
 *    under the key `"custom.freeUsed"` — same key as iOS.
 *  - Debounced custom-creature image lookup mirrors the iOS
 *    `searchTextSubject.debounce(0.5s)` Combine pipeline.
 */
class AnimalPickerViewModel(app: Application) : AndroidViewModel(app) {

    private val userPrefs = UserPrefs(app.applicationContext)
    private val cheat = CheatState
    private val imageService = AnimalImageService

    // One hot flow of raw search-text emissions for the debounce pipeline.
    private val searchTextChanges = MutableSharedFlow<String>(extraBufferCapacity = 64)

    private val _state = MutableStateFlow(UiState())
    val state: StateFlow<UiState> = _state.asStateFlow()

    private var customFetchJob: Job? = null

    init {
        // Hydrate persistent first-run flags.
        viewModelScope.launch {
            val hintCount = userPrefs.getInt(KEY_HINT_SHOWN, 0)
            val freeUsed = userPrefs.getBool(KEY_FREE_USED, false)
            _state.update { it.copy(hintShowCount = hintCount, customFreeUsed = freeUsed) }
        }

        // Debounced custom-creature info fetch — only fires when no built-in matches.
        wireCustomLookupPipeline()
    }

    @OptIn(FlowPreview::class)
    private fun wireCustomLookupPipeline() {
        viewModelScope.launch {
            searchTextChanges
                .debounce(500)
                .onEach { raw ->
                    val trimmed = raw.trim()
                    if (trimmed.isEmpty()) return@onEach
                    val hasMatches = Animals.all.any {
                        it.name.contains(trimmed, ignoreCase = true)
                    }
                    if (!hasMatches) fetchCustomAnimalInfo(trimmed)
                }
                .collect { /* consume */ }
        }
    }

    // ---- Search ---------------------------------------------------------

    fun onSearchTextChanged(text: String) {
        _state.update {
            val reset = text.trim().isEmpty()
            it.copy(
                searchText = text,
                customEmoji = if (reset) DEFAULT_EMOJI else it.customEmoji,
                customCategory = if (reset) AnimalCategory.LAND else it.customCategory,
                customColor = if (reset) DEFAULT_COLOR else it.customColor,
                customImageUrl = if (reset) null else it.customImageUrl,
            )
        }
        searchTextChanges.tryEmit(text)
    }

    fun clearSearch() = onSearchTextChanged("")

    // ---- Category / environment / arena toggle ---------------------------

    fun selectCategory(category: AnimalCategory) {
        _state.update { it.copy(selectedCategory = category) }
    }

    fun selectEnvironment(env: BattleEnvironment) {
        _state.update { it.copy(chosenEnvironment = env) }
    }

    fun setArenaEffectsEnabled(enabled: Boolean) {
        _state.update { it.copy(arenaEffectsEnabled = enabled) }
    }

    // ---- Selection ------------------------------------------------------

    /** Mirrors iOS `select(_:)` — prefers empty slot 1, then slot 2, else
     *  replaces slot 2. Never duplicates the same animal into both slots. */
    fun select(animal: Animal) {
        _state.update { s ->
            when {
                s.fighter1 == null -> s.copy(fighter1 = animal)
                s.fighter2 == null && animal != s.fighter1 -> s.copy(fighter2 = animal)
                animal != s.fighter1 -> s.copy(fighter2 = animal)
                else -> s
            }
        }
    }

    /** Convenience used by the custom-creature buttons — selects & clears the
     *  search box so the grid is immediately usable for the second pick. */
    fun selectAnimal(animal: Animal) {
        select(animal)
        onSearchTextChanged("")
    }

    fun clearSlot(slot: Int) {
        _state.update {
            when (slot) {
                1 -> it.copy(fighter1 = null)
                2 -> it.copy(fighter2 = null)
                else -> it
            }
        }
    }

    fun toggleAnimal(animal: Animal) {
        val s = _state.value
        when {
            s.fighter1?.id == animal.id -> clearSlot(1)
            s.fighter2?.id == animal.id -> clearSlot(2)
            else -> select(animal)
        }
    }

    fun reset() {
        _state.update {
            UiState(
                hintShowCount = it.hintShowCount,
                customFreeUsed = it.customFreeUsed,
            )
        }
    }

    // ---- Hint bookkeeping ----------------------------------------------

    /** Called by the view after the hint has been displayed once. Bumps the
     *  visible-count persisted counter; hint disappears at 3. */
    fun markHintShown() {
        viewModelScope.launch {
            val next = _state.value.hintShowCount + 1
            userPrefs.setInt(KEY_HINT_SHOWN, next)
            _state.update { it.copy(hintShowCount = next) }
        }
    }

    // ---- Dual-path custom creature -------------------------------------

    /** User tapped the BIG free button (first-use path). Marks `custom.freeUsed`
     *  true and selects the creature. Returns the animal that was selected. */
    fun claimFreeCustom(animal: Animal): Animal {
        viewModelScope.launch {
            userPrefs.setBool(KEY_FREE_USED, true)
            _state.update { it.copy(customFreeUsed = true) }
        }
        selectAnimal(animal)
        return animal
    }

    /** Coin path — caller already called CoinStore.spend(cost). We just select. */
    fun onCoinsSpentForCustom(animal: Animal) = selectAnimal(animal)

    /** Ad path — caller shows a rewarded ad via [AdManager] and calls this with
     *  the grant result. */
    fun onAdGrantedForCustom(animal: Animal, granted: Boolean) {
        if (granted) selectAnimal(animal)
    }

    // ---- Image lookup --------------------------------------------------

    private fun fetchCustomAnimalInfo(name: String) {
        // Only one in-flight lookup at a time; latest wins.
        customFetchJob?.cancel()
        if (!ContentFilter.isAppropriate(name)) return

        customFetchJob = viewModelScope.launch {
            coroutineScope {
                val infoDeferred = async { imageService.fetchAnimalInfo(name) }
                val imageDeferred = async { imageService.imageURL(name) }
                val info = infoDeferred.await()
                val imageUrl = imageDeferred.await()

                // Bail if the user has kept typing past this fetch.
                val current = _state.value.searchText.trim()
                if (current != name) return@coroutineScope

                _state.update {
                    it.copy(
                        customEmoji = info.emoji,
                        customCategory = info.category,
                        customColor = info.pixelColor,
                        customImageUrl = imageUrl,
                    )
                }
            }
        }
    }

    // ---- Derived helpers read by the screen -----------------------------

    /** Filters the roster by (category, search, Olympus visibility). */
    fun filteredAnimals(s: UiState = _state.value): List<Animal> {
        val olympusVisible = cheat.olympusUnlocked.value || s.isOlympusUnlockedPersisted
        return Animals.all.filter { animal ->
            if (animal.category == AnimalCategory.OLYMPUS && !olympusVisible) return@filter false
            val categoryMatch =
                s.selectedCategory == AnimalCategory.ALL || animal.category == s.selectedCategory
            val searchMatch =
                s.searchText.isBlank() || animal.name.contains(s.searchText, ignoreCase = true)
            categoryMatch && searchMatch
        }
    }

    /** The iOS `lockedAnimal` computed prop — a known creature the user
     *  hasn't unlocked whose name is an exact (case-insensitive) match. */
    fun lockedAnimal(s: UiState = _state.value): Animal? {
        val trimmed = s.searchText.trim()
        if (trimmed.isEmpty() || filteredAnimals(s).isNotEmpty()) return null
        return Animals.all.firstOrNull { it.name.equals(trimmed, ignoreCase = true) }
    }

    /** The iOS `customAnimal` computed prop. */
    fun customAnimal(s: UiState = _state.value): Animal? {
        val trimmed = s.searchText.trim()
        if (trimmed.isEmpty()) return null
        if (filteredAnimals(s).isNotEmpty()) return null
        if (lockedAnimal(s) != null) return null
        if (!ContentFilter.isAppropriate(trimmed)) return null
        return Animal(
            id = trimmed.lowercase().replace(" ", "_"),
            name = trimmed.replaceFirstChar { it.uppercase() },
            emoji = s.customEmoji,
            category = s.customCategory,
            pixelColor = s.customColor,
            size = 3,
            isCustom = true,
            imageUrl = s.customImageUrl,
        )
    }

    companion object {
        private const val KEY_HINT_SHOWN = "custom.hintShown"
        private const val KEY_FREE_USED = "custom.freeUsed"
        private const val DEFAULT_EMOJI = "🐾"
        private const val DEFAULT_COLOR = "#888888"
    }
}

/** Immutable state rendered by [AnimalPickerScreen]. */
data class UiState(
    val fighter1: Animal? = null,
    val fighter2: Animal? = null,
    val searchText: String = "",
    val selectedCategory: AnimalCategory = AnimalCategory.ALL,
    val chosenEnvironment: BattleEnvironment = BattleEnvironment.GRASSLAND,
    val arenaEffectsEnabled: Boolean = false,

    // Custom creature lookup state
    val customEmoji: String = "🐾",
    val customCategory: AnimalCategory = AnimalCategory.LAND,
    val customColor: String = "#888888",
    val customImageUrl: String? = null,

    // Persistent first-run flags (hydrated from DataStore)
    val hintShowCount: Int = 0,
    val customFreeUsed: Boolean = false,

    // Settings snapshot (hydrated from UserSettings observer)
    val isFantasyUnlockedPersisted: Boolean = false,
    val isPrehistoricUnlockedPersisted: Boolean = false,
    val isMythicUnlockedPersisted: Boolean = false,
    val isOlympusUnlockedPersisted: Boolean = false,
) {
    val canFight: Boolean get() = fighter1 != null && fighter2 != null
}

File: ios/WhoWouldWin/ViewModels/AnimalPickerViewModel.swift
import Foundation
import Combine

final class AnimalPickerViewModel: ObservableObject {
    @Published var fighter1: Animal? = nil
    @Published var fighter2: Animal? = nil
    @Published var searchText: String = "" {
        didSet {
            if searchText.trimmingCharacters(in: .whitespacesAndNewlines)
                != oldValue.trimmingCharacters(in: .whitespacesAndNewlines) {
                customAnimalID = "custom_" + UUID().uuidString.lowercased()
                    .replacingOccurrences(of: "-", with: "_")
            }
            // Reset emoji when search changes so we don't show stale info
            if searchText.trimmingCharacters(in: .whitespaces).isEmpty {
                customAnimalEmoji = "🐾"
                customAnimalCategory = .land
                customAnimalColor = "#888888"
                customAnimalImageURL = nil
            }
        }
    }
    @Published var selectedCategory: AnimalCategory = .all
    @Published var selectedEnvironment: BattleEnvironment = .grassland
    @Published var arenaEffectsEnabled: Bool = false

    // Custom animal published state
    @Published var customAnimalEmoji: String = "🐾"
    @Published var customAnimalCategory: AnimalCategory = .land
    @Published var customAnimalColor: String = "#888888"
    /// Best available image URL for the custom animal (Wikipedia or Pollinations).
    /// nil while loading; always set once fetchCustomAnimalInfo() completes.
    @Published var customAnimalImageURL: URL? = nil

    // A random identifier prevents child-entered names from becoming durable
    // identifiers in requests, logs, or database rows.
    private var customAnimalID = "custom_" + UUID().uuidString.lowercased()
        .replacingOccurrences(of: "-", with: "_")

    // MARK: - Computed

    var filteredAnimals: [Animal] {
        Animals.all.filter { animal in
            // Olympus gods: visible via cheat code OR legitimate unlock
            if animal.category == .olympus
                && !CheatState.shared.olympusUnlocked
                && !UserSettings.shared.isOlympusUnlocked { return false }
            let categoryMatch = selectedCategory == .all || animal.category == selectedCategory
            let searchMatch = searchText.isEmpty || animal.name.localizedCaseInsensitiveContains(searchText)
            return categoryMatch && searchMatch
        }
    }

    var canFight: Bool {
        fighter1 != nil && fighter2 != nil
    }

    /// Returns a known-but-locked animal if the search text matches one the user hasn't unlocked.
    /// Used to show an "unlock required" prompt instead of treating it as a custom animal.
    var lockedAnimal: Animal? {
        let trimmed = searchText.trimmingCharacters(in: .whitespaces)
        guard !trimmed.isEmpty, filteredAnimals.isEmpty else { return nil }
        return Animals.all.first { $0.name.localizedCaseInsensitiveCompare(trimmed) == .orderedSame }
    }

    var customAnimal: Animal? {
        guard !searchText.trimmingCharacters(in: .whitespaces).isEmpty,
              filteredAnimals.isEmpty,
              lockedAnimal == nil else { return nil }
        // Cap the name length — unbounded names blow out battle headers,
        // bracket cards, and share-card layouts (which render synchronously
        // and can't reflow).
        let name = String(searchText.trimmingCharacters(in: .whitespaces).prefix(24))
        guard ContentFilter.isAppropriate(name) else { return nil }
        return Animal(
            id: customAnimalID,
            name: name.capitalized,
            emoji: customAnimalEmoji,
            category: customAnimalCategory,
            pixelColor: customAnimalColor,
            size: 3,
            isCustom: true,
            imageURL: customAnimalImageURL
        )
    }

    // MARK: - Selection

    /// Convenience alias used by the custom-animal button in AnimalPickerView.
    /// Clears search text after selection so the grid is immediately ready for the next pick.
    func selectAnimal(_ animal: Animal) {
        select(animal)
        searchText = ""
    }

    /// Select an animal into the next available slot.
    /// - Slot 1 empty → fill slot 1
    /// - Slot 1 filled, slot 2 empty → fill slot 2
    /// - Both filled → replace slot 2 (most recently filled)
    /// - Prevents selecting the same animal in both slots
    func select(_ animal: Animal) {
        if fighter1 == nil {
            fighter1 = animal
        } else if fighter2 == nil {
            // Don't allow the same animal in both slots
            guard animal != fighter1 else { return }
            fighter2 = animal
        } else {
            // Both filled — replace slot 2, but don't duplicate slot 1
            guard animal != fighter1 else { return }
            fighter2 = animal
        }
    }

    /// King-of-the-hill: keep the winner in slot 1 and queue a fresh, unlocked
    /// opponent the player isn't already fighting.
    func setupNextChallenger(winner: Animal) {
        fighter1 = winner
        fighter2 = randomOpponent(excluding: winner)
        searchText = ""
        selectedEnvironment = .grassland
        arenaEffectsEnabled = false
    }

    /// A random unlocked opponent, never the given animal. Always returns
    /// something (the free packs guarantee a non-empty pool).
    private func randomOpponent(excluding animal: Animal) -> Animal {
        let unlocked = Animals.all.filter {
            $0.id != animal.id && !$0.isCustom && UserSettings.shared.isAvailable($0)
        }
        if let pick = unlocked.randomElement() { return pick }
        return Animals.all.first { $0.id != animal.id && !$0.isCustom } ?? animal
    }

    func clear(_ slot: Int) {
        switch slot {
        case 1:
            fighter1 = nil
        case 2:
            fighter2 = nil
        default:
            break
        }
    }

    func reset() {
        fighter1 = nil
        fighter2 = nil
        searchText = ""
        selectedCategory = .all
        selectedEnvironment = .grassland
        customAnimalEmoji = "🐾"
        customAnimalCategory = .land
        customAnimalColor = "#888888"
        customAnimalImageURL = nil
    }
}
